dali/operators/reader/loader/file_loader.cc

```diff
@@ -24,34 +25,53 @@
 namespace dali {
 
 
-inline void assemble_file_list(const std::string& path, const std::string& curr_entry, int label,
-                        std::vector<std::pair<std::string, int>> *file_label_pairs) {
+inline void assemble_file_list(const std::string& path, const std::string& curr_entry,
+                               const std::string& filter, std::vector<std::string>* file_list) {
   std::string curr_dir_path = path + "/" + curr_entry;
   DIR *dir = opendir(curr_dir_path.c_str());
 
   struct dirent *entry;
 
-  while ((entry = readdir(dir))) {
-    std::string full_path = curr_dir_path + "/" + std::string{entry->d_name};
+  if (filter.empty()) {
+    while ((entry = readdir(dir))) {
+      std::string full_path = curr_dir_path + "/" + std::string{entry->d_name};
+
 #ifdef _DIRENT_HAVE_D_TYPE
     /*
      * we support only regular files and symlinks, if FS returns DT_UNKNOWN
      * it doesn't mean anything and let us validate filename itself
      */
-    if (entry->d_type != DT_REG && entry->d_type != DT_LNK &&
-        entry->d_type != DT_UNKNOWN) {
-      continue;
-    }
+      if (entry->d_type != DT_REG && entry->d_type != DT_LNK &&
+         entry->d_type != DT_UNKNOWN) {
+        continue;
+      }
 #endif
-    std::string rel_path = curr_entry + "/" + std::string{entry->d_name};
-    if (HasKnownExtension(std::string(entry->d_name))) {
-      file_label_pairs->push_back(std::make_pair(rel_path, label));
+      std::string rel_path = curr_entry + "/" + std::string{entry->d_name};
+      if (HasKnownExtension(std::string(entry->d_name))) {
+         file_list->push_back(rel_path);
+      }
+    }
+  } else {
+    // use glob to do the file search
+    glob_t pglob;
+    std::string pattern = curr_dir_path + '/' + filter;
+    glob(pattern.c_str(), GLOB_TILDE, NULL, &pglob);
+
+    // iterate through the matched files
+    for (unsigned int count = 0; count < pglob.gl_pathc; ++count) {
+      std::string match(pglob.gl_pathv[count]);
+      std::string rel_path = curr_entry + "/" + match.substr(match.find_last_of("/")+1);
+      file_list->push_back(rel_path);
     }
+    // clean up
+    globfree(&pglob);
   }
   closedir(dir);
 }
 
-vector<std::pair<string, int>> filesystem::traverse_directories(const std::string& file_root) {
+
+vector<std::string> filesystem::traverse_directories(const std::string& file_root,
+                                                     std::string filter) {
```

Comment:
```suggestion
                                                     const std::string &filter) {
```
The argument is not moved elsewhere inside the function, so a string passed by value constitudes an unnecessary copy.